src/zenml/integrations/kubernetes/orchestrators/kubernetes_orchestrator.py

```diff
@@ -0,0 +1,384 @@
+#  permissions and limitations under the License.
+
+# Parts of the `prepare_or_run_pipeline()` method of this file are
+# inspired by the kubernetes dag runner implementation of tfx
+
+from typing import TYPE_CHECKING, Any, ClassVar, Dict, List, Optional, Tuple
+
+from kubernetes import client as k8s_client
+from kubernetes import config as k8s_config
+from tfx.proto.orchestration.pipeline_pb2 import Pipeline as Pb2Pipeline
+
+from zenml.enums import StackComponentType
+from zenml.environment import Environment
+from zenml.integrations.kubernetes import KUBERNETES_ORCHESTRATOR_FLAVOR
+from zenml.integrations.kubernetes.orchestrators import kube_utils
+from zenml.integrations.kubernetes.orchestrators.kubernetes_orchestrator_entrypoint_configuration import (
+    KubernetesOrchestratorEntrypointConfiguration,
+)
+from zenml.integrations.kubernetes.orchestrators.manifest_utils import (
+    build_cron_job_manifest,
+    build_pod_manifest,
+)
+from zenml.logger import get_logger
+from zenml.orchestrators import BaseOrchestrator
+from zenml.repository import Repository
+from zenml.stack import StackValidator
+from zenml.utils.docker_utils import get_image_digest
+from zenml.utils.source_utils import get_source_root_path
+
+if TYPE_CHECKING:
+    from zenml.pipelines.base_pipeline import BasePipeline
+    from zenml.runtime_configuration import RuntimeConfiguration
+    from zenml.stack import Stack
+    from zenml.steps import BaseStep
+
+logger = get_logger(__name__)
+
+
+class KubernetesOrchestrator(BaseOrchestrator):
+    """Orchestrator responsible for running pipelines using Kubernetes.
+
+    Attributes:
+        custom_docker_base_image_name: Name of a docker image that should be
+            used as the base for the image that will be run on k8s pods. If no
+            custom image is given, a basic image of the active ZenML version
+            will be used. **Note**: This image needs to have ZenML installed,
+            otherwise the pipeline execution will fail. For that reason, you
+            might want to extend the ZenML docker images found here:
+            https://hub.docker.com/r/zenmldocker/zenml/
+        kubernetes_context: Optional name of a kubernetes context to run
+            pipelines in. If not set, the current active context will be used.
+            You can find the active context by running `kubectl config
+            current-context`.
+        kubernetes_namespace: Name of the kubernetes namespace to be used.
+            If not provided, `default` namespace will be used.
+        synchronous: If `True`, running a pipeline using this orchestrator will
+            block until all steps finished running on Kubernetes.
+        skip_context_checks: If `True`, don't validate the kubernetes context.
+            This is mainly useful for unit testing.
+    """
+
+    custom_docker_base_image_name: Optional[str] = None
+    kubernetes_context: Optional[str] = None
+    kubernetes_namespace: str = "default"
```

Comment:
Yeah that makes sense, I'm also using a custom namespace in my own stack setup. 

The only thing to take care of is here is to change the kubernetes metadata store accordingly, since they need to be in the same namespace to work. In particular, if we remove the default value of `kubernetes_namespace` in the metadata store as discussed above and at the same time change the default value of the orchestrator to `zenml`, then I see the potential for users to be confused about how to set them up together. 

Personally, I'm leaning towards changing both default values to `zenml` and only have `deployment_name` as a required field for the metadata store to force users to reflect about its usage.

Reply:
Yeah I like that, default both namespaces to the same value and have the `deployment_name` required without a default value.
One helpful addition I could see in that case:
During stack validation in the `KubernetesMetadataStore` or/and `KubernetesOrchestrator`, we could check if the other component exists and fail if the namespaces don't match. Do you think that makes sense or is it too specific?

Reply:
I think that would make sense if the components were coupled, but you can use the orchestrator with other metadata stores, or the metadata store with other orchestrators, so IMO that's too specific.

Reply:
I changed both namespace defaults to `zenml` and added a custom pydantic `root_validator` to check the `deployment_name`.